Automattic/wp-calypso
# Checkout: Migrate checkout data store to createReduxStore

## Proposed Changes

The `registerStore()` function in the `@wordpress/data` package is deprecated. This PR replaces its use in the checkout data store with the recommended `createReduxStore()` function. This will give us better TypeScript support amongst other things.

The most notable difference is that we must pass the store itself to `useSelect()` and `useDispatch()` which means that we can no longer create the store on-the-fly with a React hook inside checkout. Instead we must initialize the store as a side-effect of importing the store file. To preserve some of the original behavior, this adds a new `reset()` action to the store which will reset the store's data to its original values. However, since the store was already a singleton, this PR does not call the `reset()` method anywhere except tests. Hopefully that won't cause any issues.

Part of https://github.com/Automattic/wp-calypso/issues/74399

## Testing Instructions

Automated tests should cover everything. However, it's probably a good idea to make sure that the store updates as expected. For the most part the store is used to hold contact/billing fields inside checkout, so you can edit those fields and make sure that they change as you type. It may also be a good idea to add a domain to your cart and verify that the domain fields also work.

<img width="582" alt="Screenshot 2023-03-15 at 10 47 30 AM" src="https://user-images.githubusercontent.com/2036909/225346704-61238c86-40a8-4109-a104-f8efe31f3e3b.png">

<img width="596" alt="Screenshot 2023-03-15 at 10 50 15 AM" src="https://user-images.githubusercontent.com/2036909/225347264-e5425a03-f528-4bfa-953f-8546962cc498.png">
